### PR TITLE
Fix #1280: Modify tomcat interceptor

### LINKED ISSUE #1280: RuntimeContext error when in errorHandler
Please answer these questions before submitting your issue.

- Why do you submit this issue?
- [ ] Question or discussion
- [x] Bug
- [ ] Requirement
- [ ] Feature or performance improvement

___
### Question
- What do you want to know?

___
### Bug
- Which version of SkyWalking, OS and JRE?
beta2
- Which company or project?

- What happen?
If possible, provide a way for reproducing the error. e.g. demo application, component version.

in springboot when cross `org.springframework.boot.autoconfigure.web.BasicErrorController.error(BasicErrorController.java)` this error occur
```
ERROR 2018-05-29 18:40:15 InstMethodsInter :  class[class org.springframework.boot.autoconfigure.web.BasicErrorController] before method[error] intercept failure
java.lang.IllegalStateException: No active context
	at org.apache.skywalking.apm.agent.core.context.ContextManager.getRuntimeContext(ContextManager.java:204)
	at org.apache.skywalking.apm.plugin.spring.mvc.commons.interceptor.AbstractMethodInterceptor.beforeMethod(AbstractMethodInterceptor.java:57)
	at org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstMethodsInter.intercept(InstMethodsInter.java:82)
	at org.springframework.boot.autoconfigure.web.BasicErrorController.error(BasicErrorController.java)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest$original$wJpQRycO(InvocableHandlerMethod.java:133)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest$original$wJpQRycO$accessor$QgzG0ZjS(InvocableHandlerMethod.java)
	at org.springframework.web.method.support.InvocableHandlerMethod$auxiliary$MPcIgNtw.call(Unknown Source)
	at org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstMethodsInter.intercept(InstMethodsInter.java:93)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:97)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:827)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:738)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:635)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:742)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:728)
	at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:469)
	at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:392)
	at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:311)
	at org.apache.catalina.core.StandardHostValve.custom(StandardHostValve.java:395)
	at org.apache.catalina.core.StandardHostValve.status(StandardHostValve.java:254)
	at org.apache.catalina.core.StandardHostValve.throwable(StandardHostValve.java:349)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:175)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:803)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1459)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
```
___
### Requirement or improvement
- Please describe about your requirements or improvement suggestions.

### PR DESCRIPTION
**Pull request by ascrutae**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [x] Bug fix
- [ ] New feature provided
- [ ] Improve performance

- Related issues
#1280

### CODE REVIEW

**Comment by candyleer** [comment]
@ascrutae @wu-sheng I have tried this solution, but this will trigger other problem

**Comment by wu-sheng** [comment]
@candyleer Could you be specific about how? Or trigger what?

**Comment by candyleer** [comment]
wait a moment  ,i will show you the reason

**Comment by candyleer** [comment]
in the class `org.apache.skywalking.apm.plugin.spring.mvc.commons.interceptor.AbstractMethodInterceptor`
* if more than one entry span are created,entry span will merge
* so in `beforeMethod`,the forward error request will override the request which user request.
so
```
 @Override
    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
        MethodInterceptResult result) throws Throwable {
    
        EnhanceRequireObjectCache pathMappingCache = (EnhanceRequireObjectCache)objInst.getSkyWalkingDynamicField();
        String requestURL = pathMappingCache.findPathMapping(method);
        if (requestURL == null) {
            requestURL = getRequestURL(method);
            pathMappingCache.addPathMapping(method, requestURL);
            requestURL = pathMappingCache.findPathMapping(method);
        }

        String hystrixIsolateStrategy = (String)ContextManager.getRuntimeContext().get(ISOLATE_STRATEGY_KEY_IN_RUNNING_CONTEXT);
        HttpServletRequest request = (HttpServletRequest)ContextManager.getRuntimeContext().get(REQUEST_KEY_IN_RUNTIME_CONTEXT);

        if (hystrixIsolateStrategy != null) {
            ContextManager.createLocalSpan(requestURL);
        } else if (request != null) {
            ContextCarrier contextCarrier = new ContextCarrier();
            CarrierItem next = contextCarrier.items();
            while (next.hasNext()) {
                next = next.next();
                next.setHeadValue(request.getHeader(next.getHeadKey()));
            }

            AbstractSpan span = ContextManager.createEntrySpan(requestURL, contextCarrier);
            Tags.URL.set(span, request.getRequestURL().toString());
            Tags.HTTP.METHOD.set(span, request.getMethod());
            span.setComponent(ComponentsDefine.SPRING_MVC_ANNOTATION);
            SpanLayer.asHttp(span);
        }
    }
```
the code in this forward error situation will create two entry span and  they are in the same thread,so  span and tag info will be overide

**Comment by ascrutae** [comment]
That's right. The event is the new local span that i said.

**Comment by candyleer** [comment]
good idea

**Comment by wu-sheng** [comment]
> That's right. The event is the new local span that i said.

I am a little confusing, do you say forward event of http MVC handler, or even of tracing?

Are these going to proposal?
- For a simple MVC request, it stays as same as before.
- If trigger forward, such as error handling, then each forward step will be a local span?

**Comment by wu-sheng** [comment]
I am saying, creating span for `forward` is high payload for agent. It is not necessary. Please consider
1. Use RuntimeContext to notify MVC plugins to either create an entrySpan, or just create an existed EntrySpan's event.
1. Keep `forward` and new URL as event params only.

**Comment by ascrutae** [comment]
@wu-sheng  @candyleer  I had change the code. and now  the spring mvc plugin will not create the entry span when  every forward even happened.

File: apm-sniffer/apm-sdk-plugin/tomcat-7.x-8.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/tomcat78x/define/TomcatInstrumentation.java [hunk 1]
```diff
@@ -37,7 +37,7 @@
     /**
      * Enhance class.
      */
-    private static final String ENHANCE_CLASS = "org.apache.catalina.core.StandardWrapperValve";
+    private static final String ENHANCE_CLASS = "org.apache.catalina.core.StandardHostValve";
```
Comment (candyleer) [inline apm-sniffer/apm-sdk-plugin/tomcat-7.x-8.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/tomcat78x/define/TomcatInstrumentation.java]: `exception` method may not exist in `org.apache.catalina.core.StandardHostValve`

Comment (ascrutae) [reply apm-sniffer/apm-sdk-plugin/tomcat-7.x-8.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/tomcat78x/define/TomcatInstrumentation.java]: Done!

**Comment by ascrutae** [comment]
I had test this PR in the following scenarios:

**Scenario One**:
1. visit `http://localhost:8080/test3` by using browser
2. the `/test3` mapping method throw an runtime exception

And here is the segment:
```
spans {
  parentSpanId: -1
  startTime: 1528521133425
  endTime: 1528521133644
  operationName: "/test3"
  spanLayer: Http
  componentId: 14
  isError: true
  tags {
    key: "url"
    value: "http://localhost:8080/test3"
  }
  tags {
    key: "http.method"
    value: "GET"
  }
}
```

**Scenario Two**:
1. visit `http://localhost:8080/test1` by using browser
2. The `/test1` mapping method  will forward to the URL `http://localhost:8080/test2` 
3. The `/test2` mapping method set value to redis  

And Here is the segment 
```
spans {
  spanId: 1
  parentSpanId: 0
  startTime: 1528520530757
  endTime: 1528520530769
  operationName: "Jedis/set"
  peer: "null"
  spanType: Exit
  spanLayer: Cache
  componentId: 30
  tags {
    key: "db.type"
    value: "Redis"
  }
  tags {
    key: "db.statement"
    value: "set 1"
  }
}
spans {
  parentSpanId: -1
  spanId: 0
  startTime: 1528520530426
  endTime: 1528520530792
  operationName: "/test1"
  spanLayer: Http
  componentId: 14
  tags {
    key: "url"
    value: "http://localhost:8080/test1"
  }
  tags {
    key: "http.method"
    value: "GET"
  }
}
```

**Comment by wu-sheng** [comment]
@ascrutae Where is the `forward` event?

**Comment by ascrutae** [comment]
I had add the forward event , Here is the new span like:

**Scenario One**:
```
spans {
  parentSpanId: -1
  spanId: 0
  startTime: 1528616625010
  endTime: 1528616625112
  operationName: "/test3"
  spanLayer: Http
  componentId: 14
  isError: true
  tags {
    key: "url"
    value: "http://localhost:8080/test3"
  }
  tags {
    key: "http.method"
    value: "GET"
  }
  logs {
    time: 1528616625040
    data {
      key: "forward-url"
      value: "/error"
    }
  }
}
```

**Scenario Two**:
```
spans {
  parentSpanId: 0
  spanId: 1
  startTime: 1528616245107
  endTime: 1528616245120
  operationName: "Jedis/set"
  peer: "null"
  spanType: Exit
  spanLayer: Cache
  componentId: 30
  tags {
    key: "db.type"
    value: "Redis"
  }
  tags {
    key: "db.statement"
    value: "set 1"
  }
}
spans {
  parentSpanId: -1
  spanId: 0
  startTime: 1528616244359
  endTime: 1528616245190
  operationName: "/test1"
  spanLayer: Http
  componentId: 14
  tags {
    key: "url"
    value: "http://localhost:8080/test1"
  }
  tags {
    key: "http.method"
    value: "GET"
  }
  logs {
    time: 1528616244762
    data {
      key: "forward-url"
      value: "/test2"
    }
  }
}
```

**Comment by BlackEragon** [comment]
I have meet an error with " No active context" in this place,Is there a problem with my usage?

`ERROR 2018-07-04 17:33:41 InstMethodsInter :  class[class com.hpe.msp.admin.web.api.ApiController] after method[findApiListByCondition] intercept failure 
java.lang.IllegalStateException: No active context
	at org.apache.skywalking.apm.agent.core.context.ContextManager.getRuntimeContext(ContextManager.java:204)
	at org.apache.skywalking.apm.plugin.spring.mvc.commons.interceptor.AbstractMethodInterceptor.afterMethod(AbstractMethodInterceptor.java:104)
	at org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstMethodsInter.intercept(InstMethodsInter.java:105)
	at com.hpe.msp.admin.web.api.ApiController.findApiListByCondition(ApiController.java)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)

`

**Comment by liudantop** [comment]
I have a issue, in scenario two:
1. have a API "/test_forward"
2.API "/test_forward" mapping method will forward to the URL "/error_proto"

I can not get tracing data.

debug agent workflow, I find that activeSpanStack is not empty in TracingContext, becuase  step6 direct return, then stackDepth = 1  in step7 stopSpan, :
1: TomcatInvokeInterceptor(tomcat plugin) > before method
	2: AbstractMethodInterceptor(spring plugin) > before method
 		3: ForwardInterceptor(tomcat) > before(set forward flag)
			4: AbstractMethodInterceptor > before method(because forward flag then return )
			5: AbstractMethodInterceptor > before method(because forward flag then return )
	6: AbstractMethodInterceptor > after method(because forward flag then return )
7: TomcatInvokeInterceptor(tomcat plugin) > after method

**Comment by wu-sheng** [comment]
Please submit this to new issue if you could provide demo codes to reproduce the issue.

**Pull request merged**